scheduler/services/resource/base.py

```diff
@@ -0,0 +1,584 @@
+                    while missing_row_date < row_date:
+                        # Make sure there is an entry and append to it to avoid overwriting anything already present.
+                        date_set = self._resources[site].setdefault(missing_row_date, set())
+                        self._resources[site][missing_row_date] = date_set | copy(self._resources[site][prev_row_date])
+                        missing_row_date += FileBasedResourceService._day
+
+                # Get or create date_set for the date, and append new resources from table, ignoring blank entries.
+                date_set = self._resources[site].setdefault(row_date, set())
+                new_entries = {self.lookup_resource(r) for r in c(row[1:], site) if r}
+                self._resources[site][row_date] = date_set | new_entries
+
+                # Advance the previous row date where data was defined.
+                prev_row_date = row_date
+
+        if type(data_source) is str:
+            with open(os.path.join(self._path, data_source)) as f:
+                _process_file(f)
+        else:
+            data_str = data_source.getvalue().decode()
+            f = StringIO(data_str)
+            _process_file(f)
+
+    @staticmethod
+    def _split_program_ids(text: str) -> Set[ProgramID]:
+        """
+        Given a text list that should contain comma-separated program IDs, separate them
+        and return the set of them.
+        """
+        return set(ProgramID(prog_id.strip()) for prog_id in text.strip().split(','))
+
+    @staticmethod
+    def _add_dates_to_dict(prog_dict: Dict[ProgramID, Set[date]], text: str, add_date: date) -> None:
+        """
+        Given a program dictionary, parse the text following it into program IDs and add the
+        specified date to each set associated with each program ID.
+        """
+        for prog_id in FileBasedResourceService._split_program_ids(text):
+            date_set = prog_dict.setdefault(prog_id, set())
+            date_set.add(add_date)
+
+    def _load_spreadsheet(self, file_source: Union[str, BytesIO],
+                          from_gdrive: bool = False) -> None:
+        """
+        Process an Excel spreadsheet containing instrument, mode, and LGS information.
+
+        The Excel spreadsheets have information available for every date, so we do not have to concern ourselves
+        as in the _load_csv file above.
+        """
+
+        def none_to_str(value) -> str:
+            return '' if value is None else value
+
+        if not file_source:
+            raise ValueError('file_source cannot be empty')
+
+        # filename = os.path.join(self._path, OcsResourceService._SITE_CONFIG_FILE)
+        if from_gdrive:
+            logger.info('Retrieving site configuration file from Google Drive...')
+            try:
+                GoogleDriveDownloader.download_file_from_google_drive(file_id=FileBasedResourceService._SITE_CONFIG_GOOGLE_ID,
```

Comment:
Just a note to both of us: at some point, we should standardize this and do this for all the OCS files.